### DISCUSSION TITLE
How to make sure the user is logged in before serving other routes?

### DISCUSSION THREAD

**Discussion by teenjuna** [maxcountryman/axum-login#59] [Q&A]
Maybe the title is a bit confusing. In my app, every session has an associated user. A user can be either `guest` or `normal`. For every new session, a new `guest` user is created. The later part is what brings me here!

My router looks like this:
```rust
let router = Router::new()
    .route("/", get(index))
    .layer(middleware::from_fn_with_state(state.clone(), assign_user))
    .layer(auth_layer)
    .layer(session_layer)
    .with_state(state);
```

Index route is just for debug:
```rust
async fn index(
    session: ReadableSession,
    auth: AuthContext<User, UserStore, UserRole>,
) -> Json<(Session, Option<User>)> {
    tracing::debug!("Serving index page");
    Json((Session::from(session.clone()), auth.current_user))
}
```

And here is the `assign_user` middleware that creates `guest` user if session doesn't have a user:
```rust
async fn assign_user<B>(
    mut auth: AuthContext<User, UserStore, UserRole>,
    State(db): State<Db>,
    request: Request<B>,
    next: Next<B>,
) -> Result<Response, StatusCode> {
    if auth.current_user.is_some() {
        return Ok(next.run(request).await);
    }
    let user: User = db.create(User::table()).await.unwrap();
    auth.login(&user).await.unwrap();
    tracing::debug!("New user was created for a session: {:?}", user);
    Ok(next.run(request).await)
}
```

The problem: in my logs I see that first the new user is created in the middleware, then the index route is served, and only then the updated session (with the new user) is stored. The result of this is that JSON output on index route shows the session record which contains `_auth_id` and `_user_id` fields, but the `auth.current_user` is `None`. Only after I refresh the page the user is loaded. This is problematic since I need to have the user on the first run for the frontend on my actual app. @maxcountryman, do you have any idea how that can be achieved?

**Comment by teenjuna** [comment]
Although an answer might be useful for future readers, it's not relevant for me anymore. I went ahead and removed both `axum-sessions` and `axum-login` and used `axum-extra`'s `SignedCookieJar` + middleware fn instead. Got exactly what I need with a lot less code! Not to say something bad about those libs, of course. Just to throw the fact that sometimes one can do what he wants using simple things.

**Reply by maxcountryman** [reply]
> Although an answer might be useful for future readers, it's not relevant for me anymore. I went ahead and removed both `axum-sessions` and `axum-login` and used `axum-extra`'s `SignedCookieJar` + middleware fn instead. Got exactly what I need with a lot less code! Not to say something bad about those libs, of course. Just to throw the fact that sometimes one can do what he wants using simple things.

Glad you found a solution that worked for you. I'll leave the discussion open for other folks.
